walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,120 @@
+      for (String optional: optionalGuests) {
+        if (guests.contains(optional)) {
+          badTimes.add(eventTime);
+        }
+      }
+      for (String person: meetingGuests) {
+        if (guests.contains(person)) {
+          fixConflicts(potentialTimes, eventTime);
+        }
+      }
+    }
+
+    ArrayList<TimeRange> copyOfGoodTimes = new ArrayList<TimeRange>(potentialTimes);
+    HashMap<TimeRange, Integer> someOptionals = new HashMap<>();
+ 
+    for (TimeRange badTime: badTimes) {
+      boolean accounted = false;
+      ListIterator<TimeRange> litr = potentialTimes.listIterator();
+      while (litr.hasNext()) {
+        TimeRange time = litr.next();
+        if (badTime.overlaps(time)) {
+          litr.remove();
+          someOptionals.put(time, 1);
+          if (time.start() < badTime.start()) {
+            TimeRange frontBlock = TimeRange.fromStartEnd(time.start(), badTime.start(), false);
+            litr.add(frontBlock);
+          }
+          if (badTime.end() < time.end()) {
+            TimeRange backBlock = TimeRange.fromStartEnd(badTime.end(), time.end(), false);
+            litr.add(backBlock);
+          }
+        }
+        accounted = true; 
+      }
+      if (!accounted) {
+          for (TimeRange deletedTime: someOptionals.keySet()) {
+              if (badTime.overlaps(deletedTime)) {
+                  int count = someOptionals.get(deletedTime);
+                  someOptionals.replace(deletedTime, count + 1);
+              }
+          }
+      }
+    }
+
+    if (meetingGuests.isEmpty()) {
+      copyOfGoodTimes = potentialTimes;
+    }
+
+    ListIterator<TimeRange> litr = potentialTimes.listIterator();
+    while (litr.hasNext()) {
+      TimeRange time = litr.next(); 
+      if (meetingDuration > time.duration()) {
+        litr.remove();
+      }
+    } 
+
+    if (potentialTimes.isEmpty() && !badTimes.isEmpty()) {
+      potentialTimes = copyOfGoodTimes;
+      ArrayList<TimeRange> leastClashes = new ArrayList<>();
+      if (!someOptionals.isEmpty() && !meetingGuests.isEmpty()) {
+          Collection<Integer> values = someOptionals.values();
+          int min = Collections.min(values); 
+          for (TimeRange time: someOptionals.keySet()) {
+              if (someOptionals.get(time) == min) {
+                  leastClashes.add(time);
+              }
+          }
+      }
+      if (!leastClashes.isEmpty()) potentialTimes = leastClashes;
+    }
+
+    times = potentialTimes;    
+    return times;
+  }
+
+  public void fixConflicts(ArrayList<TimeRange> potentialTimes, TimeRange clashTime) {
+    ListIterator<TimeRange> litr = potentialTimes.listIterator();
+    while (litr.hasNext()) {
+      TimeRange time = litr.next();
+      if (clashTime.overlaps(time)) {
+        litr.remove();
+        if (time.start() < clashTime.start()) {
+          TimeRange frontBlock = TimeRange.fromStartEnd(time.start(), clashTime.start(), false);
+          litr.add(frontBlock);
```

Comment:
Are `frontBlock` and `backBlock` the open time in the potential time before/after the clashing time? Does that mean this function removes the clashing time from any potential times?

Oh! that's we start with All Day at the top. That's clever!